File: libView/src/main/java/org/waynezhou/libView/transition/ViewStep.java
package org.waynezhou.libView.transition;

import android.animation.Animator;
import android.animation.PropertyValuesHolder;
import android.animation.ValueAnimator;
import android.view.View;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public class ViewStep<TValueHolder> {

    public static class Builder<TValueHolder>  {
        private final ViewTransition.Builder<TValueHolder> host;
        private final ViewTransition.ViewGetter viewGetter;
        private final ViewTransition.Condition condition;
        protected Builder(ViewTransition.Builder<TValueHolder> host, ViewTransition.ViewGetter viewGetter) {
            this(host, viewGetter, null);
        }
        protected Builder(ViewTransition.Builder<TValueHolder> host, ViewTransition.ViewGetter viewGetter, ViewTransition.Condition condition) {
            this.host = host;
            this.viewGetter = viewGetter;
            this.condition = condition;
        }


        protected final List<ViewPropStep<TValueHolder>> propSteps = new ArrayList<>();

        public ViewPropStep.Builder<TValueHolder> let(String prop) {
            return new ViewPropStep.Builder<>(this, prop);
        }

        protected final List<ViewPropFix<TValueHolder>> propFixes = new ArrayList<>();

        public ViewPropFix.Builder<TValueHolder> fix(String prop) {
            return new ViewPropFix.Builder<>(this, prop);
        }

        public ViewStep.Builder<TValueHolder> preBuild(ViewPreBuild<TValueHolder> preBuild){
            preBuild.build(this);
            return this;
        }

        protected ViewAction preAction = v->{};
        public Builder<TValueHolder> pre(ViewAction action) {
            preAction = action;
            return this;
        }

        public ViewTransition.Builder<TValueHolder> $_endAddStep() {
            if(condition!=null){
                host.conditionalViewSteps.add(new ViewTransition.ConditionalViewStep<>(condition, new ViewStep<>(viewGetter, propSteps, propFixes, preAction)));
            }else{
                host.viewSteps.add(new ViewStep<>(viewGetter, propSteps, propFixes, preAction));
            }
            return host;
        }
    }

    private final ViewTransition.ViewGetter viewGetter;
    private final List<ViewPropStep<TValueHolder>> propSteps;
    protected final List<ViewPropFix<TValueHolder>> propFixes;
    protected final ViewAction preAction;
    private ViewStep(ViewTransition.ViewGetter viewGetter, List<ViewPropStep<TValueHolder>> propSteps, List<ViewPropFix<TValueHolder>> fixes, ViewAction preAction) {
        this.viewGetter = viewGetter;
        this.propSteps = Collections.unmodifiableList(propSteps);
        this.propFixes = Collections.unmodifiableList(fixes);
        this.preAction = preAction;
    }

    @SuppressWarnings("unchecked")
    public Animator createAnimator(TValueHolder holder, ViewAnimatorArgs args) {
        final List<PropertyValuesHolder> holders = new ArrayList<>();
        for (int i = 0; i < propSteps.size(); i++) {
            holders.add(propSteps.get(i).createHolder(viewGetter, holder));
        }
        final PropertyValuesHolder[] holderArr = holders.toArray(new PropertyValuesHolder[0]);
        ValueAnimator ret = new ValueAnimator();
        ret.setDuration(args.duration);
        ret.setInterpolator(args.interpolator);
        ret.setValues(holderArr);
        ret.addUpdateListener(animator -> {
            final View view = viewGetter.get();
            for (int i = 0; i < holders.size(); i++) {
                final PropertyValuesHolder pvHolder = holders.get(i);
                final LayoutTransitionPropertyBridge bridge = Objects.requireNonNull(LayoutTransitionPropertyBridges.bridges.get(pvHolder.getPropertyName()));
                bridge.set(view, (float) animator.getAnimatedValue(pvHolder.getPropertyName()));
                /**
                 LayoutTransitionPropertyBridge.TranslationX.update(view);
                 LayoutTransitionPropertyBridge.TranslationY.update(view);
                 LayoutTransitionPropertyBridge.ScaleX.update(view);
                 LayoutTransitionPropertyBridge.ScaleY.update(view);
                 */
            }
            for (int i = 0; i < propFixes.size(); i++) {
                final ViewPropFix<TValueHolder> fix = propFixes.get(i);
                final LayoutTransitionPropertyBridge bridge = Objects.requireNonNull(LayoutTransitionPropertyBridges.bridges.get(fix.propName));
                ValueGetter getter = fix.getter;
                if (getter instanceof ValueGetter.FromView) {
                    ((ValueGetter.FromView) getter).setView(viewGetter.get());
                }
                if (getter instanceof ValueGetter.FromValueHolder) {
                    ((ValueGetter.FromValueHolder<TValueHolder>) getter).setHolder(holder);
                }
                bridge.set(view, getter.get());

            }
            view.requestLayout();

        });
        ret.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
                preAction.execute(viewGetter.get());
            }

            @Override
            public void onAnimationEnd(Animator animation) {

            }

            @Override
            public void onAnimationCancel(Animator animation) {

            }

            @Override
            public void onAnimationRepeat(Animator animation) {

            }
        });
        return ret;
    }

    @SuppressWarnings("unchecked")
    public ViewStepExecutor createExecutor(TValueHolder holder) {
        final List<ViewPropStep.ViewPropFinalStep> finalSteps = new ArrayList<>();
        for (int i = 0; i < propSteps.size(); i++) {
            finalSteps.add(propSteps.get(i).createFinalStep(viewGetter, holder));
        }
        return (withRequestLayout) -> {
            final View view = viewGetter.get();
            preAction.execute(view);
            for (int i = 0; i < finalSteps.size(); i++) {
                final ViewPropStep.ViewPropFinalStep finalStep = finalSteps.get(i);
                final LayoutTransitionPropertyBridge bridge = Objects.requireNonNull(LayoutTransitionPropertyBridges.bridges.get(finalStep.propName));
                bridge.set(view, finalStep.value);
            }
            for (int i = 0; i < propFixes.size(); i++) {
                final ViewPropFix<TValueHolder> fix = propFixes.get(i);
                final LayoutTransitionPropertyBridge bridge = Objects.requireNonNull(LayoutTransitionPropertyBridges.bridges.get(fix.propName));
                ValueGetter getter = fix.getter;
                if (getter instanceof ValueGetter.FromView) {
                    ((ValueGetter.FromView) getter).setView(viewGetter.get());
                }
                if (getter instanceof ValueGetter.FromValueHolder) {
                    ((ValueGetter.FromValueHolder<TValueHolder>) getter).setHolder(holder);
                }
                bridge.set(view, getter.get());
            }
            if (withRequestLayout) view.requestLayout();
        };
    }

    public ViewStep<TValueHolder> transpose() {
        final List<ViewPropStep<TValueHolder>> transposedPropSteps = new ArrayList<>();
        for (ViewPropStep<TValueHolder> vpStep : this.propSteps) {
            transposedPropSteps.add(vpStep.transpose());
        }
        final List<ViewPropFix<TValueHolder>> transposedPropFixes = new ArrayList<>();
        for (ViewPropFix<TValueHolder> vpFix : this.propFixes) {
            transposedPropFixes.add(vpFix.transpose());
        }
        return new ViewStep<>(this.viewGetter, transposedPropSteps, transposedPropFixes, this.preAction);
    }

    @FunctionalInterface
    public interface ViewStepExecutor {
        void execute(boolean withRequestLayout);
    }

    @FunctionalInterface
    public interface ViewAction {
        void execute(View view);
    }
    @FunctionalInterface
    public interface ViewPreBuild<TValueHolder> {
        void build(ViewStep.Builder<TValueHolder> it);
    }
}

File: libViewKt/src/main/java/org/waynezhou/libviewkt/view_transition/ViewStep.kt
package org.waynezhou.libviewkt.view_transition

import android.animation.Animator
import android.animation.ValueAnimator
import android.view.View

/** Use getter to store layout of a view. */
class ViewStep<TValueHolder : Any> private constructor(
    private val viewGetter: () -> View,
    private val propSteps: List<ViewPropStep<TValueHolder>>,
    private val propFixes: List<ViewPropFix<TValueHolder>>,
    private val preAction: (View) -> Unit,
    private val postAction: (View) -> Unit
) {
    /** Builder of [ViewTransition]. */
    class Builder<TValueHolder : Any, TView : View> private constructor(
        private val host: ViewTransition.Builder<TValueHolder>,
        private val viewGetter: () -> TView,
        internal val propSteps: MutableList<ViewPropStep<TValueHolder>>,
        internal val propFixes: MutableList<ViewPropFix<TValueHolder>>,
        private val condition: (() -> Boolean)?,
        private var preAction: (TView) -> Unit,
        private var postAction: (TView) -> Unit
    ) {
        /** Empty instance builder Constructor. */
        constructor(host: ViewTransition.Builder<TValueHolder>, getter: () -> TView) : this(
            host,
            getter,
            mutableListOf(),
            mutableListOf(),
            null,
            {}, {})

        /** Empty instance with condition builder constructor. */
        constructor(
            host: ViewTransition.Builder<TValueHolder>,
            condition: () -> Boolean,
            getter: () -> TView
        ) : this(
            host,
            getter,
            mutableListOf(),
            mutableListOf(),
            condition,
            {}, {})

        /** Copy Instance Builder Constructor. */
        constructor(
            host: ViewTransition.Builder<TValueHolder>,
            other: ViewStep<TValueHolder>
        ) : this(
            host,
            other.viewGetter as () -> TView,
            other.propSteps.toMutableList(),
            other.propFixes.toMutableList(),
            null,
            { other.preAction },
            { other.postAction }
        )

        fun let(prop: String): ViewPropStep.Builder<TValueHolder, TView> {
            return ViewPropStep.Builder(this, prop)
        }

        fun preAction(action: TView.() -> Unit): Builder<TValueHolder, TView> {
            preAction = action
            return this
        }

        fun postAction(action: TView.() -> Unit): Builder<TValueHolder, TView> {
            postAction = action
            return this
        }

        fun endAddStep(): ViewTransition.Builder<TValueHolder> {
            if (condition == null)
                host.viewSteps.add(
                    ViewStep(
                        viewGetter,
                        propSteps,
                        propFixes,
                        preAction as (View) -> Unit,
                        postAction as (View) -> Unit
                    )
                )
            else
                host.conditionalViewSteps.add(
                    ConditionalViewStep(
                        condition,
                        ViewStep(
                            viewGetter,
                            propSteps,
                            propFixes,
                            preAction as (View) -> Unit,
                            postAction as (View) -> Unit
                        )
                    )
                )
            return host
        }
    }

    internal fun createAnimator(holder: TValueHolder, args: ViewAnimatorArgs): Animator {

        val holders = propSteps.map { it.createHolder(viewGetter, holder) }

        val holderArr = holders.toTypedArray()

        val ret = ValueAnimator()
        ret.duration = args.duration
        ret.interpolator = args.interpolator
        ret.setValues(*holderArr)
        ret.addUpdateListener { animator: ValueAnimator ->
            val view: View = viewGetter()
            holders.forEach {
                propBridges[it.propertyName]?.let { bridge ->
                    bridge.set(view, animator.getAnimatedValue(it.propertyName) as Float)
                    return@forEach
                }
                throw NullPointerException()
            }

            propFixes.forEach {
                propBridges[it.propName]?.let { bridge ->
                    when (val getter = it.getter) {
                        is ValueGetter.FromView -> bridge.set(view, getter.getFrom(view))
                        is ValueGetter.FromHolder<*> -> bridge.set(view, getter.getFrom(holder))
                    }
                    return@forEach
                }
                throw NullPointerException()
            }

            view.requestLayout()
        }
        ret.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                preAction(viewGetter())
            }

            override fun onAnimationEnd(animation: Animator) {
                postAction(viewGetter())
            }

            override fun onAnimationCancel(animation: Animator) {}
            override fun onAnimationRepeat(animation: Animator) {}
        })
        return ret
    }

    internal fun createExecutor(holder: TValueHolder): ViewStepExecutor {
        val finalSteps = propSteps.map {
            it.createFinalStep(viewGetter, holder)
        }

        return ViewStepExecutor { withRequestLayout: Boolean ->
            val view = viewGetter()
            preAction(view)

            finalSteps.forEach {
                propBridges[it.propName]?.let { bridge ->
                    bridge.set(view, it.value)
                    return@forEach
                }
                throw NullPointerException()
            }

            propFixes.forEach {
                propBridges[it.propName]?.let { bridge ->
                    when (val getter = it.getter) {
                        is ValueGetter.FromView -> bridge.set(view, getter.getFrom(view))
                        is ValueGetter.FromHolder<*> -> bridge.set(view, getter.getFrom(holder))
                    }
                    return@forEach
                }
                throw NullPointerException()
            }
            if (withRequestLayout) view.requestLayout()
        }
    }

    internal fun transpose(): ViewStep<TValueHolder> {
        val transposedPropSteps = propSteps.map { it.transpose() }

        val transposedPropFixes = propFixes.map { it.transpose() }

        return ViewStep(
            viewGetter,
            transposedPropSteps,
            transposedPropFixes,
            this.preAction,
            this.postAction
        )
    }

    internal fun interface ViewStepExecutor {
        fun execute(withRequestLayout: Boolean)
    }
}


internal class ConditionalViewStep<TValueHolder : Any> internal constructor(
    internal val condition: () -> Boolean,
    internal val viewStep: ViewStep<TValueHolder>
)
